pilot/pkg/model/virtualservice.go

```diff
@@ -0,0 +1,349 @@
+)
+
+func mergeVirtualServicesIfNeeded(vServices []Config) (out []Config) {
+	out = make([]Config, 0, len(vServices))
+	delegatesMap := map[string]Config{}
+	// root virtualservices with delegate
+	var rootVses []Config
+
+	// 1. classify virtualservices
+	for _, vs := range vServices {
+		rule := vs.Spec.(*networking.VirtualService)
+		// it is delegate, add it to the indexer cache
+		if len(rule.Hosts) == 0 {
+			delegatesMap[key(vs.Name, vs.Namespace)] = vs
+			continue
+		}
+
+		// root vs
+		if isRootVs(rule) {
+			rootVses = append(rootVses, vs)
+			continue
+		}
+
+		// the others are normal vs without delegate
+		out = append(out, vs)
+	}
+
+	// If `PILOT_ENABLE_VIRTUAL_SERVICE_DELEGATE` feature disabled,
+	// filter out invalid vs(root or delegate), this can happen after enable -> disable
+	if !features.EnableVirtualServiceDelegate.Get() {
+		return
+	}
+
+	// 2. merge delegates and root
+	for _, root := range rootVses {
+		rootVs := root.Spec.(*networking.VirtualService)
+		mergedRoutes := []*networking.HTTPRoute{}
+		for _, route := range rootVs.Http {
+			// it is root vs with delegate
+			if route.Delegate != nil {
+				delegate, ok := delegatesMap[key(route.Delegate.Name, route.Delegate.Namespace)]
+				if !ok {
+					log.Debugf("delegate virtual service %s/%s of %s/%s not found",
+						route.Delegate.Namespace, route.Delegate.Name, root.Namespace, root.Name)
+					// delegate not found, ignore only the current HTTP route
+					continue
+				}
+				// DeepCopy to prevent mutate the original delegate, it can conflict
+				// when multiple routes delegate to one single VS.
+				copiedDelegate := delegate.DeepCopy()
+				vs := copiedDelegate.Spec.(*networking.VirtualService)
+				merged := mergeHTTPRoute(route, vs.Http)
+				mergedRoutes = append(mergedRoutes, merged...)
+			} else {
+				mergedRoutes = append(mergedRoutes, route)
+			}
+		}
+		rootVs.Http = mergedRoutes
+		if log.DebugEnabled() {
+			jsonm := &jsonpb.Marshaler{Indent: "   "}
+			vsString, _ := jsonm.MarshalToString(rootVs)
+			log.Infof("merged virtualService: %s", vsString)
+		}
+		out = append(out, root)
+	}
+
+	return
+}
+
+// merge root's route with delegate's and the merged route number equals the delegate's.
+// if there is a conflict with root, the route is ignored
+func mergeHTTPRoute(root *networking.HTTPRoute, delegate []*networking.HTTPRoute) []*networking.HTTPRoute {
+	root.Delegate = nil
+
+	out := make([]*networking.HTTPRoute, 0, len(delegate))
+	for _, subRoute := range delegate {
+		// suppose there are N1 match conditions in root, N2 match conditions in delegate
+		// if match condition of N2 is a subset of anyone in N1, this is a valid matching conditions
+		merged, conflict := mergeHTTPMatchRequests(root.Match, subRoute.Match)
+		if conflict {
+			log.Debugf("HTTPMatchRequests conflict: root root %s, delegate root %s", root.Name, subRoute.Name)
+			continue
+		}
+		subRoute.Match = merged
+
+		if subRoute.Name == "" {
```

Comment:
can we factor all of this logic in to a separate function and add tests (including field count if possible, see below)?